File: leetcode/js/198.house-robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
    if (nums.length == 0) return 0;
    if (nums.length == 1) return nums[0];
    if (nums.length == 2) return Math.max(nums[0], nums[1]);

    let dp = [nums[0], Math.max(nums[0], nums[1])];

    for (let i = 2; i < nums.length; i++) {
        const element = nums[i];
        dp.push(Math.max(element + dp[i - 2], dp[i - 1]));
    }

    return dp[dp.length - 1];

};

File: leetcode/kt/198.house-robber.kt
import kotlin.math.max

class Solution_v1 {
    fun rob(nums: IntArray): Int {
        if(nums.isEmpty())
            return 0
        if(nums.size == 1)
            return nums[0]

        val arr = IntArray(nums.size);
        arr[0] = nums[0]
        arr[1] = max(nums[0],nums[1])
        for(i in 2 until nums.size){
            arr[i] = max(nums[i] + arr[i-2],arr[i-1])
        }
        return arr.last()
    }
}

class Solution_v2 {
    fun rob(nums: IntArray): Int {
        if (nums.isEmpty())
            return 0
        if (nums.size == 1)
            return nums[0]

        var preOpt = nums[0]
        var curOpt = max(nums[0], nums[1])
        for (i in 2 until nums.size) {
            val nextOpt = max(nums[i] + preOpt, curOpt)
            preOpt = curOpt;
            curOpt = nextOpt;
        }
        return curOpt
    }
}